From 34fdda895db367f225f49d94a55a8f04aa89842f Mon Sep 17 00:00:00 2001
From: Troy Frew <fuzzy_bot@arenaga.me>
Date: Tue, 15 Nov 2016 09:07:51 -0500
Subject: [PATCH] Fixup ProtocolConstants


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index b3d1c2bb..ad9aa920 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -125,6 +125,16 @@ public class ProtocolConstants
         SUPPORTED_VERSION_IDS = supportedVersionIds.build();
     }
 
+    public static final boolean isBeforeOrEq(int before, int other)
+    {
+            return before <= other;
+    }
+
+    public static final boolean isAfterOrEq(int after, int other)
+    {
+            return after >= other;
+    }
+
     public enum Direction
     {
 
-- 
2.47.1

