From b5c843ea9c67d816e7fd64f89aced24fa356d3b0 Mon Sep 17 00:00:00 2001
From: Tux <write@imaginarycode.com>
Date: Sun, 14 Feb 2016 01:03:27 -0500
Subject: [PATCH] Reduce max log size to 8MB and don't append to existing log
 files


diff --git a/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java b/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java
index 9598fff..2a2e0ed 100644
--- a/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java
+++ b/log/src/main/java/net/md_5/bungee/log/BungeeLogger.java
@@ -27,7 +27,7 @@ public class BungeeLogger extends Logger
 
         try
         {
-            FileHandler fileHandler = new FileHandler( filePattern, 1 << 24, 8, true );
+            FileHandler fileHandler = new FileHandler( filePattern, 1 << 23, 8, false );
             fileHandler.setFormatter( formatter );
             addHandler( fileHandler );
 
-- 
2.7.4

