From 8b9b91fa1167cef1ae3ee78dd06e94125ebb64dc Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@techcable.net>
Date: Mon, 6 Jun 2016 13:36:10 -0600
Subject: [PATCH] Don't send KICK packets while in HANDSHAKE state


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index f18ab46..88bc825 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -551,7 +551,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
             @Override
             public void run()
             {
-                if ( thisState != State.STATUS && thisState != State.PING )
+                if ( thisState != State.STATUS && thisState != State.PING && thisState != State.HANDSHAKE ) // Waterfall: Don't kick during handshake
                 {
                     unsafe().sendPacket( new Kick( ComponentSerializer.toString( reason ) ) );
                 }
-- 
2.10.0

