From a9437edffbbe4bbee2b0521cf894a413c845a3cc Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Sat, 10 Jun 2017 20:56:02 +0100
Subject: [PATCH] Improve outdated build message


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index 39fc55f9..c81a4430 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -53,11 +53,11 @@ public class BungeeCordLauncher
             deadline.add( Calendar.WEEK_OF_YEAR, -8 );
             if ( buildDate.before( deadline.getTime() ) )
             {
-                System.err.println( "*** Warning, this build is outdated ***" );
-                System.err.println( "*** Please download a new build from https://ci.destroystokyo.com/job/Waterfall/ ***" );
-                System.err.println( "*** You will get NO support regarding this build ***" );
-                System.err.println( "*** Server will start in 10 seconds ***" );
-                Thread.sleep( TimeUnit.SECONDS.toMillis( 10 ) );
+                System.err.println( "*** Hey! This build is potentially outdated :( ***" );
+                System.err.println( "*** Please check for a new build from https://papermc.io/downloads ***" );
+                System.err.println( "*** Should this build be outdated, you will get NO support for it. ***" );
+                //System.err.println( "*** Server will start in 10 seconds ***" );
+                //Thread.sleep( TimeUnit.SECONDS.toMillis( 10 ) );
             }
         }
 
-- 
2.38.1

