From b01f3decf7099e6df46436c73e3c9538cfad54cd Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Sat, 10 Jun 2017 20:56:02 +0100
Subject: [PATCH] Improve outdated build message


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index 4ba0ced5..4c119b9d 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -48,9 +48,9 @@ public class BungeeCordLauncher
             deadline.add( Calendar.WEEK_OF_YEAR, -8 );
             if ( buildDate.before( deadline.getTime() ) )
             {
-                System.err.println( "*** Warning, this build is outdated ***" );
-                System.err.println( "*** Please download a new build from https://ci.destroystokyo.com/job/Waterfall/ ***" );
-                System.err.println( "*** You will get NO support regarding this build ***" );
+                System.err.println( "*** Hey! This build is potentially outdated :( ***" );
+                System.err.println( "*** Please check for a new build from https://papermc.io/ci/job/Waterfall/ ***" );
+                System.err.println( "*** Should this build be outdated, you will get NO support for it. ***" );
                 System.err.println( "*** Server will start in 10 seconds ***" );
                 Thread.sleep( TimeUnit.SECONDS.toMillis( 10 ) );
             }
-- 
2.24.0

