From 1c564bb488c8c51dc801442fce67f76ff68208fe Mon Sep 17 00:00:00 2001
From: joserobjr <jose.rob.jr@gmail.com>
Date: Wed, 8 Jun 2016 11:32:15 -0700
Subject: [PATCH] Add support for Minecraft 1.10


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 59aa5d4..92d53e4 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -129,7 +129,8 @@ public enum Protocol
                     PlayerListHeaderFooter.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x47 ),
                     map( ProtocolConstants.MINECRAFT_1_9, 0x48 ),
-                    map(ProtocolConstants.MINECRAFT_1_9_4, 0x47 )
+                    map(ProtocolConstants.MINECRAFT_1_9_4, 0x47 ),
+                    map(ProtocolConstants.MINECRAFT_1_10, 0x47 )
             );
 
             TO_SERVER.registerPacket(
@@ -258,7 +259,8 @@ public enum Protocol
             ));
             linkedProtocols.put( ProtocolConstants.MINECRAFT_1_9, Arrays.asList(ProtocolConstants.MINECRAFT_1_9_1,
                 ProtocolConstants.MINECRAFT_1_9_2,
-                ProtocolConstants.MINECRAFT_1_9_4
+                ProtocolConstants.MINECRAFT_1_9_4,
+                ProtocolConstants.MINECRAFT_1_10
             ));
         }
 
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 2501512..80250de 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -11,15 +11,18 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_9_1 = 108;
     public static final int MINECRAFT_1_9_2 = 109;
     public static final int MINECRAFT_1_9_4 = 110;
+    public static final int MINECRAFT_1_10 = 210;
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.8.x",
-            "1.9.x"
+            "1.9.x",
+            "1.10"
     );
     public static final List<Integer> SUPPORTED_VERSION_IDS = Arrays.asList(ProtocolConstants.MINECRAFT_1_8,
             ProtocolConstants.MINECRAFT_1_9,
             ProtocolConstants.MINECRAFT_1_9_1,
             ProtocolConstants.MINECRAFT_1_9_2,
-            ProtocolConstants.MINECRAFT_1_9_4
+            ProtocolConstants.MINECRAFT_1_9_4,
+            ProtocolConstants.MINECRAFT_1_10
     );
 
     public enum Direction
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
index 9ba9633..947beb3 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
@@ -32,6 +32,7 @@ public abstract class EntityMap
             case ProtocolConstants.MINECRAFT_1_9_2:
                 return EntityMap_1_9.INSTANCE;
             case ProtocolConstants.MINECRAFT_1_9_4:
+            case ProtocolConstants.MINECRAFT_1_10:
                 return EntityMap_1_9_4.INSTANCE;
         }
         throw new RuntimeException( "Version " + version + " has no entity map" );
-- 
2.8.3

