From a2614c692d4251c501768ba11f6fe2566274cad2 Mon Sep 17 00:00:00 2001
From: Jamie Mansfield <dev@jamierocks.uk>
Date: Sun, 5 Aug 2018 13:54:39 +0100
Subject: [PATCH] temporary: Fix broken upstream build


diff --git a/api/src/main/java/net/md_5/bungee/api/ServerConnectRequest.java b/api/src/main/java/net/md_5/bungee/api/ServerConnectRequest.java
index 578546dd..33d2f446 100644
--- a/api/src/main/java/net/md_5/bungee/api/ServerConnectRequest.java
+++ b/api/src/main/java/net/md_5/bungee/api/ServerConnectRequest.java
@@ -59,7 +59,7 @@ public class ServerConnectRequest
     /**
      * Timeout in milliseconds for request.
      */
-    @Builder.Default
+    @lombok.Builder.Default
     private final int connectTimeout = 5000; // TODO: Configurable
     /**
      * Should the player be attempted to connect to the next server in their
-- 
2.18.0

