From c2a02e4a3c6b76b529f4de6bdf8a381892579ef2 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 21 Jul 2018 17:14:39 +0100
Subject: [PATCH] 1.13 protocol support


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index a8ac332a..a248c7db 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -267,42 +267,36 @@ public enum Protocol
         {
             TO_CLIENT.registerPacket(
                     LoginPayloadRequest.class,
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x00 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x04 )
             );
             TO_CLIENT.registerPacket(
                     Kick.class,
-                    map( ProtocolConstants.MINECRAFT_1_8, 0x00 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x01 )
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
             TO_CLIENT.registerPacket(
                     EncryptionRequest.class,
-                    map( ProtocolConstants.MINECRAFT_1_8, 0x01 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
             );
             TO_CLIENT.registerPacket(
                     LoginSuccess.class,
-                    map( ProtocolConstants.MINECRAFT_1_8, 0x02 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x03 )
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x02 )
             );
             TO_CLIENT.registerPacket(
                     SetCompression.class,
-                    map( ProtocolConstants.MINECRAFT_1_8, 0x03 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x04 )
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x03 )
             );
 
             TO_SERVER.registerPacket(
                     LoginPayloadResponse.class,
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x00 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
             );
             TO_SERVER.registerPacket(
                     LoginRequest.class,
-                    map( ProtocolConstants.MINECRAFT_1_8, 0x00 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x01 )
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x00 )
             );
             TO_SERVER.registerPacket(
                     EncryptionResponse.class,
-                    map( ProtocolConstants.MINECRAFT_1_8, 0x01 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
+                    map( ProtocolConstants.MINECRAFT_1_8, 0x01 )
             );
         }
     };
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index bba6cb2d..b2dc9423 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -17,7 +17,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_12 = 335;
     public static final int MINECRAFT_1_12_1 = 338;
     public static final int MINECRAFT_1_12_2 = 340;
-    public static final int MINECRAFT_1_13 = 389;
+    public static final int MINECRAFT_1_13 = 393;
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.8.x",
             "1.9.x",
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/ScoreboardObjective.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/ScoreboardObjective.java
index 6279d9f3..edcc59a1 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/ScoreboardObjective.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/ScoreboardObjective.java
@@ -1,5 +1,7 @@
 package net.md_5.bungee.protocol.packet;
 
+import net.md_5.bungee.api.chat.TextComponent; // Waterfall - 1.13
+import net.md_5.bungee.chat.ComponentSerializer; // Waterfall - 1.13
 import net.md_5.bungee.protocol.DefinedPacket;
 import io.netty.buffer.ByteBuf;
 import java.util.Locale;
@@ -35,9 +37,12 @@ public class ScoreboardObjective extends DefinedPacket
             value = readString( buf );
             if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_13 )
             {
+                String valueJson = readString( buf ); // Waterfall - 1.13
+                value = TextComponent.toLegacyText(ComponentSerializer.parse(valueJson)); // Waterfall - 1.13
                 type = HealthDisplay.values()[readVarInt( buf )];
             } else
             {
+                value = readString( buf ); // Waterfall - 1.13
                 type = HealthDisplay.fromString( readString( buf ) );
             }
         }
@@ -50,12 +55,14 @@ public class ScoreboardObjective extends DefinedPacket
         buf.writeByte( action );
         if ( action == 0 || action == 2 )
         {
-            writeString( value, buf );
             if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_13 )
             {
+                String valueJson = ComponentSerializer.toString(TextComponent.fromLegacyText(value)); // Waterfall - 1.13
+                writeString( valueJson, buf ); // Waterfall - 1.13
                 writeVarInt( type.ordinal(), buf );
             } else
             {
+                writeString( value, buf ); // Waterfall - 1.13
                 writeString( type.toString(), buf );
             }
         }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Team.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Team.java
index f93508d9..6688b287 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Team.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Team.java
@@ -1,5 +1,7 @@
 package net.md_5.bungee.protocol.packet;
 
+import net.md_5.bungee.api.chat.TextComponent;
+import net.md_5.bungee.chat.ComponentSerializer;
 import net.md_5.bungee.protocol.DefinedPacket;
 import io.netty.buffer.ByteBuf;
 import lombok.AllArgsConstructor;
@@ -46,7 +48,14 @@ public class Team extends DefinedPacket
         mode = buf.readByte();
         if ( mode == 0 || mode == 2 )
         {
-            displayName = readString( buf );
+            // Waterfall start - 1.13
+            if (protocolVersion <= ProtocolConstants.MINECRAFT_1_13) {
+                String displayNameJson = readString(buf);
+                displayName = TextComponent.toLegacyText(ComponentSerializer.parse(displayNameJson));
+            } else {
+                displayName = readString( buf );
+            }
+            // Waterfall end
             if ( protocolVersion < ProtocolConstants.MINECRAFT_1_13 )
             {
                 prefix = readString( buf );
@@ -61,8 +70,12 @@ public class Team extends DefinedPacket
             color = ( protocolVersion >= ProtocolConstants.MINECRAFT_1_13 ) ? readVarInt( buf ) : buf.readByte();
             if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_13 )
             {
-                prefix = readString( buf );
-                suffix = readString( buf );
+                // Waterfall start - 1.13
+                String prefixJson = readString( buf );
+                String suffixJson = readString( buf );
+                prefix = TextComponent.toLegacyText(ComponentSerializer.parse(prefixJson));
+                suffix = TextComponent.toLegacyText(ComponentSerializer.parse(suffixJson));
+                // Waterfall end
             }
         }
         if ( mode == 0 || mode == 3 || mode == 4 )
@@ -83,7 +96,14 @@ public class Team extends DefinedPacket
         buf.writeByte( mode );
         if ( mode == 0 || mode == 2 )
         {
-            writeString( displayName, buf );
+            // Waterfall start - 1.13
+            if (protocolVersion <= ProtocolConstants.MINECRAFT_1_13) {
+                String displayNameJson = ComponentSerializer.toString(TextComponent.fromLegacyText(displayName));
+                writeString( displayNameJson, buf );
+            } else {
+                writeString( displayName, buf );
+            }
+            // Waterfall end - 1.13
             if ( protocolVersion < ProtocolConstants.MINECRAFT_1_13 )
             {
                 writeString( prefix, buf );
@@ -98,9 +118,14 @@ public class Team extends DefinedPacket
 
             if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_13 )
             {
+                // Waterfall start - 1.13
+                String prefixJson = ComponentSerializer.toString(TextComponent.fromLegacyText(prefix));
+                String suffixJson = ComponentSerializer.toString(TextComponent.fromLegacyText(suffix));
+
                 writeVarInt( color, buf );
-                writeString( prefix, buf );
-                writeString( suffix, buf );
+                writeString( prefixJson, buf );
+                writeString( suffixJson, buf );
+                // Waterfall end - 1.13
             } else
             {
                 buf.writeByte( color );
-- 
2.18.0

