From 1f214e0f51add45580c27f75825d1516ff438827 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 30 Mar 2019 15:11:11 +0000
Subject: [PATCH] Fix upstream javadocs


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index f3bced0a..4a198ee9 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -17,27 +17,25 @@ public interface ProxyConfig
     /**
      * Time before users are disconnected due to no network activity.
      *
-     * @return timeout
+     * @return the time before users are disconnected due to no network activity.
      */
     int getTimeout();
 
     /**
      * UUID used for metrics.
-     *
-     * @return uuid
+     * @return the UUID used for metrics
      */
     String getUuid();
 
     /**
      * Set of all listeners.
-     *
-     * @return listeners
+     * @return a set of all listeners
      */
     Collection<ListenerInfo> getListeners();
 
     /**
      * Set of all servers.
-     * @return servers
+     * @return a map of registered servers
      * @deprecated The returned map may be modified concurrently by the proxy.
      *             The safe alternative is {@link #getServersCopy()}.
      */
@@ -129,14 +127,14 @@ public interface ProxyConfig
     /**
      * Does the server authenticate with Mojang.
      *
-     * @return online mode
+     * @return if the server authenticates to mojang
      */
     boolean isOnlineMode();
 
     /**
      * Whether proxy commands are logged to the proxy log.
      *
-     * @return log commands
+     * @return should the proxy log commands
      */
     boolean isLogCommands();
 
@@ -158,7 +156,7 @@ public interface ProxyConfig
     /**
      * A collection of disabled commands.
      *
-     * @return disabled commands
+     * @return a collection of disabled commands
      */
     Collection<String> getDisabledCommands();
 
@@ -181,15 +179,15 @@ public interface ProxyConfig
     /**
      * The connection throttle delay.
      *
-     * @return throttle
+     * @return the connection throttle delay
      */
     @Deprecated
     int getThrottle();
 
     /**
-     * Whether the proxy will parse IPs with spigot or not.
+     * Whether the proxy will forward the players information (UUID, IP)
      *
-     * @return ip forward
+     * @return should we forward player information
      */
     @Deprecated
     boolean isIpForward();
@@ -197,7 +195,7 @@ public interface ProxyConfig
     /**
      * The encoded favicon.
      *
-     * @return favicon
+     * @return the encoded favicon
      * @deprecated Use #getFaviconObject instead.
      */
     @Deprecated
@@ -206,7 +204,7 @@ public interface ProxyConfig
     /**
      * The favicon used for the server ping list.
      *
-     * @return favicon
+     * @return the favicon used for the server ping list
      */
     Favicon getFaviconObject();
 
-- 
2.23.0.windows.1

