From fe4c48c2acf9657709e84d43f3c686e94f86797b Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Sat, 30 Mar 2019 15:11:11 +0000
Subject: [PATCH] Fix upstream javadocs


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index ab012bff..fe46b9f3 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -17,27 +17,25 @@ public interface ProxyConfig
     /**
      * Time before users are disconnected due to no network activity.
      *
-     * @return timeout
+     * @return the time before users are disconnected due to no network activity.
      */
     int getTimeout();
 
     /**
      * UUID used for metrics.
-     *
-     * @return uuid
+     * @return the UUID used for metrics
      */
     String getUuid();
 
     /**
      * Set of all listeners.
-     *
-     * @return listeners
+     * @return a set of all listeners
      */
     Collection<ListenerInfo> getListeners();
 
     /**
      * Set of all servers.
-     * @return servers
+     * @return a map of registered servers
      * @deprecated The returned map may be modified concurrently by the proxy.
      *             The safe alternative is {@link #getServersCopy()}.
      */
@@ -129,14 +127,14 @@ public interface ProxyConfig
     /**
      * Does the server authenticate with Mojang.
      *
-     * @return online mode
+     * @return if the server authenticates to mojang
      */
     boolean isOnlineMode();
 
     /**
      * Whether proxy commands are logged to the proxy log.
      *
-     * @return log commands
+     * @return should the proxy log commands
      */
     boolean isLogCommands();
 
@@ -158,22 +156,22 @@ public interface ProxyConfig
     /**
      * A collection of disabled commands.
      *
-     * @return disabled commands
+     * @return a collection of disabled commands
      */
     Collection<String> getDisabledCommands();
 
     /**
      * The connection throttle delay.
      *
-     * @return throttle
+     * @return the connection throttle delay
      */
     @Deprecated
     int getThrottle();
 
     /**
-     * Whether the proxy will parse IPs with spigot or not
+     * Whether the proxy will forward the players information (UUID, IP)
      *
-     * @return ip forward
+     * @return should we forward player information
      */
     @Deprecated
     boolean isIpForward();
@@ -181,7 +179,7 @@ public interface ProxyConfig
     /**
      * The encoded favicon.
      *
-     * @return favicon
+     * @return the encoded favicon
      * @deprecated Use #getFaviconObject instead.
      */
     @Deprecated
@@ -190,7 +188,7 @@ public interface ProxyConfig
     /**
      * The favicon used for the server ping list.
      *
-     * @return favicon
+     * @return the favicon used for the server ping list
      */
     Favicon getFaviconObject();
 
diff --git a/api/src/main/java/net/md_5/bungee/api/scheduler/TaskScheduler.java b/api/src/main/java/net/md_5/bungee/api/scheduler/TaskScheduler.java
index 32732703..46d70184 100644
--- a/api/src/main/java/net/md_5/bungee/api/scheduler/TaskScheduler.java
+++ b/api/src/main/java/net/md_5/bungee/api/scheduler/TaskScheduler.java
@@ -85,6 +85,7 @@ public interface TaskScheduler
 
         /**
          * An executor service which underlies this scheduler.
+         * @param plugin The plugin of which to fetch an executor service for
          *
          * @param plugin owning plugin
          * @return the underlying executor service or compatible wrapper
-- 
2.25.0

