From 54ceab1968c1a8ce967d57bdd8edd409167f1464 Mon Sep 17 00:00:00 2001
From: Colin Godsey <crgodsey@gmail.com>
Date: Tue, 23 Apr 2019 11:18:07 -0600
Subject: [PATCH] TEMP 1.14 version bump


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index fb583238..a0c0b5ca 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -20,7 +20,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_13 = 393;
     public static final int MINECRAFT_1_13_1 = 401;
     public static final int MINECRAFT_1_13_2 = 404;
-    public static final int MINECRAFT_1_14 = 476;
+    public static final int MINECRAFT_1_14 = 477;
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.8.x",
             "1.9.x",
-- 
2.21.0

