From b3d3bae6ec02dfe1194e96ea8ba9f07176d3bc5b Mon Sep 17 00:00:00 2001
From: "Five (Xer)" <admin@xernium.com>
Date: Wed, 29 Apr 2020 20:52:13 +0200
Subject: [PATCH] Waterfall 1.16


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/DefinedPacket.java b/protocol/src/main/java/net/md_5/bungee/protocol/DefinedPacket.java
index e56bfccb..26513a6e 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/DefinedPacket.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/DefinedPacket.java
@@ -1,15 +1,98 @@
 package net.md_5.bungee.protocol;
 
+// Waterfall (1.16 20w21a+) imports start
 import com.google.common.base.Charsets;
 import io.netty.buffer.ByteBuf;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.UUID;
+
+import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
 import lombok.RequiredArgsConstructor;
+import net.md_5.bungee.protocol.nbt.NBTTag;
+import net.md_5.bungee.protocol.nbt.NBTTagTypes;
+// end
 
 @RequiredArgsConstructor
 public abstract class DefinedPacket
 {
+    // Waterfall (1.16 20w21a+) substitute NBT reader/ writer start
+    public static void writeNBT( NBTTag toWrite, ByteBuf buf )
+    {
+        if( toWrite == null )
+        {
+            buf.writeByte(0);
+        } else
+        {
+            try 
+            {
+                DataOutput to = new ByteBufOutputStream( buf );
+                to.writeByte( toWrite.getTagIndex() );
+                if ( toWrite.getTagIndex() != 0 ) 
+                {
+                    to.writeUTF( "" );
+                    toWrite.write( to );
+                }
+            } catch ( IOException e )
+            {
+                e.printStackTrace();
+            }
+        }
+
+    }
+
+    public static NBTTag readNBT( ByteBuf buf )
+    {
+        int indexBefore = buf.readerIndex();
+        byte startType = buf.readByte();
+        if( startType == 0 )
+        {
+            return null;
+        } else
+        {
+            buf.readerIndex( indexBefore );
+            try 
+            {
+                DataInput from = new ByteBufInputStream( buf );
+                byte type = from.readByte();
+                if ( type == 0 )
+                {
+                    return NBTTagTypes.newInstance( type );
+                } else 
+                {
+                    from.readUTF();
+                    NBTTag ret = NBTTagTypes.newInstance( type );
+                    ret.read( from );
+                    return ret;
+                }
+
+            } catch ( IOException e )
+            {
+                e.printStackTrace();
+                return null;
+            }
+        }
+    }
+
+    // Waterfall (1.16) new UUID reader/writer start
+    public static void writeUUIDIntArray(UUID u, ByteBuf buf)
+    {
+        buf.writeInt((int) (u.getMostSignificantBits() >> 32));
+        buf.writeInt((int) u.getMostSignificantBits());
+        buf.writeInt((int) (u.getLeastSignificantBits() >> 32));
+        buf.writeInt((int) u.getLeastSignificantBits());
+    }
+
+    public static UUID readUUIDIntArray(ByteBuf buf)
+    {
+        return new UUID((long) buf.readInt() << 32 | ((long) buf.readInt() & 0xFFFFFFFFL), (long) buf.readInt() << 32 | ((long) buf.readInt() & 0xFFFFFFFFL));
+    }
+    // Waterfall end
 
     public static void writeString(String s, ByteBuf buf)
     {
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index aa1515f4..6dd261aa 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -158,7 +158,8 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x42 ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x45 ),
                     map( ProtocolConstants.MINECRAFT_1_14, 0x49 ),
-                    map( ProtocolConstants.MINECRAFT_1_15, 0x4A )
+                    map( ProtocolConstants.MINECRAFT_1_15, 0x4A ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x4B ) // Waterfall (1.16: 20w12a+ 0x4A -> 0x4B)
             );
             TO_CLIENT.registerPacket(
                     ScoreboardScore.class,
@@ -169,7 +170,8 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x45 ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x48 ),
                     map( ProtocolConstants.MINECRAFT_1_14, 0x4C ),
-                    map( ProtocolConstants.MINECRAFT_1_15, 0x4D )
+                    map( ProtocolConstants.MINECRAFT_1_15, 0x4D ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x4E ) // Waterfall (1.16: 20w12a+ 0x4D -> 0x4E)
             );
             TO_CLIENT.registerPacket(
                     ScoreboardDisplay.class,
@@ -180,7 +182,8 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x3B ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x3E ),
                     map( ProtocolConstants.MINECRAFT_1_14, 0x42 ),
-                    map( ProtocolConstants.MINECRAFT_1_15, 0x43 )
+                    map( ProtocolConstants.MINECRAFT_1_15, 0x43 ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x44 ) // Waterfall (1.16: 20w12a+ 0x43 -> 0x44)
             );
             TO_CLIENT.registerPacket(
                     Team.class,
@@ -191,7 +194,8 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x44 ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x47 ),
                     map( ProtocolConstants.MINECRAFT_1_14, 0x4B ),
-                    map( ProtocolConstants.MINECRAFT_1_15, 0x4C )
+                    map( ProtocolConstants.MINECRAFT_1_15, 0x4C ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x4D ) // Waterfall (1.16: 20w12a+ 0x4C -> 0x4D)
             );
             TO_CLIENT.registerPacket(
                     PluginMessage.class,
@@ -268,7 +272,8 @@ public enum Protocol
                     map( ProtocolConstants.MINECRAFT_1_12, 0x0C ),
                     map( ProtocolConstants.MINECRAFT_1_12_1, 0x0B ),
                     map( ProtocolConstants.MINECRAFT_1_13, 0x0E ),
-                    map( ProtocolConstants.MINECRAFT_1_14, 0x0F )
+                    map( ProtocolConstants.MINECRAFT_1_14, 0x0F ),
+                    map( ProtocolConstants.MINECRAFT_1_16, 0x10 ) // Waterfall (1.16: 20w16a+ 0x0F -> 0x10)
             );
             TO_SERVER.registerPacket(
                     Chat.class,
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index cd987c8a..23d57c71 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -28,6 +28,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_15 = 573;
     public static final int MINECRAFT_1_15_1 = 575;
     public static final int MINECRAFT_1_15_2 = 578;
+    public static final int MINECRAFT_1_16 = 719; // Waterfall (1.16 - 20w22a)
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.8.x",
             "1.9.x",
@@ -36,7 +37,8 @@ public class ProtocolConstants
             "1.12.x",
             "1.13.x",
             "1.14.x",
-            "1.15.x"
+            "1.15.x",
+            "1.16" // Waterfall (1.16)
     );
     public static final List<Integer> SUPPORTED_VERSION_IDS = Arrays.asList(
             ProtocolConstants.MINECRAFT_1_8,
@@ -60,7 +62,8 @@ public class ProtocolConstants
             ProtocolConstants.MINECRAFT_1_14_4,
             ProtocolConstants.MINECRAFT_1_15,
             ProtocolConstants.MINECRAFT_1_15_1,
-            ProtocolConstants.MINECRAFT_1_15_2
+            ProtocolConstants.MINECRAFT_1_15_2,
+            ProtocolConstants.MINECRAFT_1_16 // Waterfall (1.16)
     );
 
     public static final boolean isBeforeOrEq(int before, int other)
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/nbt/CompoundTag.java b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/CompoundTag.java
new file mode 100644
index 00000000..7c289085
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/CompoundTag.java
@@ -0,0 +1,65 @@
+package net.md_5.bungee.protocol.nbt;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+ * Waterfall NBT-Library substitute module
+ * Minecraft 1.16 - 20w21a+
+ * by Xernium (Five)
+ * */
+
+public class CompoundTag extends NBTTag {
+
+    private Map<String, NBTTag> tagMap = new HashMap<String, NBTTag>();
+
+    @Override
+    public int getTagIndex()
+    {
+        return 10;
+    }
+
+    @Override
+    public void write( DataOutput to ) throws IOException
+    {
+        for( Map.Entry<String, NBTTag> entry : tagMap.entrySet() )
+        {
+            to.writeByte( entry.getValue().getTagIndex() );
+            if( entry.getValue().getTagIndex() != 0 )
+            {
+                to.writeUTF( entry.getKey() );
+                entry.getValue().write( to );
+            }
+        }
+
+        to.writeByte( 0 );
+    }
+
+    @Override
+    public void read( DataInput from ) throws IOException
+    {
+        tagMap = new HashMap<String, NBTTag>();
+        for ( byte type = from.readByte(); type != 0; type = from.readByte())
+        {
+            String key = from.readUTF();
+            NBTTag value = NBTTagTypes.newInstance( type );
+            value.read( from );
+            tagMap.put( key, value );
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        String ret = "CompoundTag: \n";
+        for( Map.Entry<String, NBTTag> entry : tagMap.entrySet() )
+        {
+           ret += "\t" + entry.getKey() + "\n";
+           ret += "\t\t" + entry.getValue().toString() + "\n";
+        }
+        return ret;
+    }
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/nbt/EndTag.java b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/EndTag.java
new file mode 100644
index 00000000..1b763a05
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/EndTag.java
@@ -0,0 +1,36 @@
+package net.md_5.bungee.protocol.nbt;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/*
+ * Waterfall NBT-Library substitute module
+ * Minecraft 1.16 - 20w21a+
+ * by Xernium (Five)
+ * */
+
+public class EndTag extends NBTTag {
+
+    @Override
+    public int getTagIndex()
+    {
+        return 0;
+    }
+
+    @Override
+    public void write(DataOutput to) throws IOException
+    {
+    }
+
+    @Override
+    public void read(DataInput from) throws IOException
+    {
+    }
+
+    @Override
+    public String toString()
+    {
+        return "EndTag\n";
+    }
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/nbt/ListTag.java b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/ListTag.java
new file mode 100644
index 00000000..66075d87
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/ListTag.java
@@ -0,0 +1,78 @@
+package net.md_5.bungee.protocol.nbt;
+
+import com.google.common.collect.Lists;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.ArrayList;
+
+/*
+ * Waterfall NBT-Library substitute module
+ * Minecraft 1.16 - 20w21a+
+ * by Xernium (Five)
+ * */
+
+
+public class ListTag extends NBTTag {
+
+    private ArrayList<NBTTag> values = new ArrayList<NBTTag>();
+    private byte tagType;
+
+    @Override
+    public int getTagIndex()
+    {
+        return 9;
+    }
+
+    @Override
+    public void write( DataOutput to ) throws IOException
+    {
+        if( values.isEmpty() )
+        {
+            to.writeByte( 0 );
+            to.writeInt( 0 );
+        } else{
+            to.writeByte( tagType );
+            to.writeInt( values.size() );
+
+            for(NBTTag iter : values)
+            {
+                iter.write(to);
+            }
+
+        }
+    }
+
+    @Override
+    public void read(DataInput from) throws IOException
+    {
+        tagType = from.readByte();
+        int len = from.readInt();
+        if( tagType == 0 && len > 0 )
+        {
+            throw new IOException( "ListTag contained filled EndTag or type for Tag was invalid" );
+        } else
+        {
+            ArrayList<NBTTag> read = Lists.newArrayListWithCapacity( len );
+            for( int i = 0; i < len; i++ )
+            {
+                NBTTag inst = NBTTagTypes.newInstance( tagType );
+                inst.read( from );
+                read.add( inst );
+            }
+            values = read;
+        }
+    }
+
+    @Override
+    public String toString()
+    {
+        String ret = "ListTag (Type: " + tagType + " : \n";
+        for( NBTTag d : values )
+        {
+           ret += "\t" + d.toString() + "\n";
+        }
+        return ret;
+    }
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/nbt/NBTTag.java b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/NBTTag.java
new file mode 100644
index 00000000..f7524c42
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/NBTTag.java
@@ -0,0 +1,23 @@
+package net.md_5.bungee.protocol.nbt;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/*
+ * Waterfall NBT-Library substitute module
+ * Minecraft 1.16 - 20w21a+
+ * by Xernium (Five)
+ * */
+
+
+public abstract class NBTTag {
+
+    public abstract int getTagIndex();
+
+    public abstract void write( DataOutput to ) throws IOException;
+
+    public abstract void read( DataInput from ) throws IOException;
+
+    public abstract String toString();
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/nbt/NBTTagTypes.java b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/NBTTagTypes.java
new file mode 100644
index 00000000..112ce81f
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/NBTTagTypes.java
@@ -0,0 +1,31 @@
+package net.md_5.bungee.protocol.nbt;
+
+import java.io.IOException;
+
+/*
+ * Waterfall NBT-Library substitute module
+ * Minecraft 1.16 - 20w21a+
+ * by Xernium (Five)
+ * */
+
+
+public class NBTTagTypes {
+
+    public static NBTTag newInstance( byte id ) throws IOException
+    {
+        switch( id )
+        {
+            case 0:
+                return new EndTag();
+            case 8:
+                return new StringTag();
+            case 9:
+                return new ListTag();
+            case 10:
+                return new CompoundTag();
+            default:
+                throw new IOException( "Unsupported NBT ID: " + id );
+        }
+    }
+
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/nbt/StringTag.java b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/StringTag.java
new file mode 100644
index 00000000..aefdd039
--- /dev/null
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/nbt/StringTag.java
@@ -0,0 +1,42 @@
+package net.md_5.bungee.protocol.nbt;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+
+/*
+ * Waterfall NBT-Library substitute module
+ * Minecraft 1.16 - 20w21a+
+ * by Xernium (Five)
+ * */
+
+public class StringTag extends NBTTag {
+
+    private String value;
+
+    @Override
+    public int getTagIndex()
+    {
+        return 8;
+    }
+
+    @Override
+    public void write( DataOutput to ) throws IOException
+    {
+        to.writeUTF( value );
+    }
+
+    @Override
+    public void read( DataInput from ) throws IOException
+    {
+        value = from.readUTF();
+    }
+
+    @Override
+    public String toString()
+    {
+        return "value: '" + value + "'\n";
+    }
+
+
+}
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Chat.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Chat.java
index 0ded6739..7bd9d1d6 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Chat.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Chat.java
@@ -9,6 +9,8 @@ import net.md_5.bungee.protocol.AbstractPacketHandler;
 import net.md_5.bungee.protocol.DefinedPacket;
 import net.md_5.bungee.protocol.ProtocolConstants;
 
+import java.util.UUID; // Waterfall (1.16: 20w21a+)
+
 @Data
 @NoArgsConstructor
 @AllArgsConstructor
@@ -18,10 +20,11 @@ public class Chat extends DefinedPacket
 
     private String message;
     private byte position;
+    private UUID sender; // Waterfall (1.16: 20w21a+)
 
     public Chat(String message)
     {
-        this( message, (byte) 0 );
+        this( message, (byte) 0, /* Waterfall (1.16: 20w21a+) */ new UUID( 0, 0 ) );
     }
 
     @Override
@@ -36,6 +39,12 @@ public class Chat extends DefinedPacket
         if ( direction == ProtocolConstants.Direction.TO_CLIENT )
         {
             position = buf.readByte();
+            // Waterfall (1.16: 20w21a+) start:
+            if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+            {
+                sender = new UUID( buf.readLong(), buf.readLong() );
+            }
+            // Waterfall end
         }
     }
 
@@ -51,6 +60,13 @@ public class Chat extends DefinedPacket
         if ( direction == ProtocolConstants.Direction.TO_CLIENT )
         {
             buf.writeByte( position );
+            // Waterfall (1.16: 20w21a+) start:
+            if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+            {
+                buf.writeLong( sender.getMostSignificantBits() );
+                buf.writeLong( sender.getLeastSignificantBits() );
+            }
+            // Waterfall end
         }
     }
 
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Commands.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Commands.java
index 2f14d074..56af8b57 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Commands.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Commands.java
@@ -547,6 +547,7 @@ public class Commands extends DefinedPacket
             PROVIDERS.put( "minecraft:entity_summon", VOID );
             PROVIDERS.put( "minecraft:dimension", VOID );
             PROVIDERS.put( "minecraft:time", VOID ); // 1.14
+            PROVIDERS.put( "minecraft:uuid", VOID ); // Waterfall (1.16: 20w17a+)
         }
 
         private static ArgumentType<?> read(String key, ByteBuf buf)
@@ -618,6 +619,7 @@ public class Commands extends DefinedPacket
             registerDummy( "minecraft:all_recipes" );
             registerDummy( "minecraft:available_sounds" );
             registerDummy( "minecraft:summonable_entities" );
+            registerDummy( "minecraft:available_biomes" ); // Waterfall (1.16: 20w06a)
         }
 
         private static void registerDummy(String name)
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Login.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Login.java
index e0cdf82e..461ad219 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Login.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Login.java
@@ -8,6 +8,7 @@ import lombok.NoArgsConstructor;
 import net.md_5.bungee.protocol.AbstractPacketHandler;
 import net.md_5.bungee.protocol.DefinedPacket;
 import net.md_5.bungee.protocol.ProtocolConstants;
+import net.md_5.bungee.protocol.nbt.NBTTag; // Waterfall (1.16: 20w21a+)
 
 @Data
 @NoArgsConstructor
@@ -26,19 +27,42 @@ public class Login extends DefinedPacket
     private int viewDistance;
     private boolean reducedDebugInfo;
     private boolean normalRespawn;
+    private boolean isDebug; // Waterfall (1.16: 20w20a+)
+    private boolean isFlat; // Waterfall (1.16: 20w20a+)
+    private String dimensionCodecName; // Waterfall (1.16: 20w21a+)
+    private NBTTag codecTag; // Waterfall (1.16: 20w21a+)
+    private String[] worlds; // Waterfall (1.16: 20w22a+)
+    private String world; // Waterfall (1.16: 20w22a+)
 
     @Override
     public void read(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion)
     {
         entityId = buf.readInt();
         gameMode = buf.readUnsignedByte();
-        if ( protocolVersion > ProtocolConstants.MINECRAFT_1_9 )
+        // Waterfall (1.16: 20w21a+) start:
+        if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
         {
-            dimension = buf.readInt();
+            // Waterfall (1.16: 20w22a) world name array
+            int worldNameArrayLen = readVarInt( buf );
+            worlds = new String[worldNameArrayLen];
+            for ( int iter = 0; iter < worldNameArrayLen; iter++ )
+            {
+                worlds[iter] = readString( buf );
+            }
+            codecTag = readNBT( buf );
+            dimensionCodecName = readString( buf );
+            world = readString( buf ); // Waterfall (1.16: 20w22a) world name
         } else
         {
-            dimension = buf.readByte();
+            if ( protocolVersion > ProtocolConstants.MINECRAFT_1_9 )
+            {
+                dimension = buf.readInt();
+            } else
+            {
+                dimension = buf.readByte();
+            }
         }
+        // Waterfall end
         if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_15 )
         {
             seed = buf.readLong();
@@ -48,7 +72,12 @@ public class Login extends DefinedPacket
             difficulty = buf.readUnsignedByte();
         }
         maxPlayers = buf.readUnsignedByte();
-        levelType = readString( buf );
+        // Waterfall (1.16: 20w20a+) start:
+        if( protocolVersion < ProtocolConstants.MINECRAFT_1_16 )
+        {
+            levelType = readString( buf );
+        }
+        // Waterfall end
         if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_14 )
         {
             viewDistance = readVarInt( buf );
@@ -61,6 +90,13 @@ public class Login extends DefinedPacket
         {
             normalRespawn = buf.readBoolean();
         }
+        // Waterfall (1.16: 20w20a+) start:
+        if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            isDebug = buf.readBoolean();
+            isFlat = buf.readBoolean();
+        }
+        // Waterfall end
     }
 
     @Override
@@ -68,13 +104,29 @@ public class Login extends DefinedPacket
     {
         buf.writeInt( entityId );
         buf.writeByte( gameMode );
-        if ( protocolVersion > ProtocolConstants.MINECRAFT_1_9 )
+        // Waterfall (1.16: 20w21a+) start:
+        if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
         {
-            buf.writeInt( dimension );
+            // Waterfall (1.16: 20w22a) world name array
+            writeVarInt( worlds.length, buf );
+            for ( int iter = 0; iter < worlds.length; iter++ )
+            {
+                writeString( worlds[iter], buf );
+            }
+            writeNBT( codecTag, buf );
+            writeString( dimensionCodecName, buf );
+            writeString( world, buf ); // Waterfall (1.16: 20w22a) world name
         } else
         {
-            buf.writeByte( dimension );
+            if ( protocolVersion > ProtocolConstants.MINECRAFT_1_9 )
+            {
+                buf.writeInt( dimension );
+            } else
+            {
+                buf.writeByte( dimension );
+            }
         }
+        // Waterfall end
         if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_15 )
         {
             buf.writeLong( seed );
@@ -84,7 +136,12 @@ public class Login extends DefinedPacket
             buf.writeByte( difficulty );
         }
         buf.writeByte( maxPlayers );
-        writeString( levelType, buf );
+        // Waterfall (1.16: 20w20a+) start:
+        if( protocolVersion < ProtocolConstants.MINECRAFT_1_16 )
+        {
+            writeString( levelType, buf );
+        }
+        // Waterfall end
         if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_14 )
         {
             writeVarInt( viewDistance, buf );
@@ -97,6 +154,13 @@ public class Login extends DefinedPacket
         {
             buf.writeBoolean( normalRespawn );
         }
+        // Waterfall (1.16: 20w20a+) start:
+        if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            buf.writeBoolean( isDebug );
+            buf.writeBoolean( isFlat );
+        }
+        // Waterfall end
     }
 
     @Override
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginSuccess.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginSuccess.java
index 2aa5ad58..aa50f70d 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginSuccess.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/LoginSuccess.java
@@ -7,6 +7,9 @@ import lombok.EqualsAndHashCode;
 import lombok.NoArgsConstructor;
 import net.md_5.bungee.protocol.AbstractPacketHandler;
 import net.md_5.bungee.protocol.DefinedPacket;
+import net.md_5.bungee.protocol.ProtocolConstants;
+
+import java.util.UUID;
 
 @Data
 @NoArgsConstructor
@@ -17,18 +20,33 @@ public class LoginSuccess extends DefinedPacket
 
     private String uuid;
     private String username;
+    private int protocolVersion; // Waterfall (1.16: 20w12a+)
 
     @Override
     public void read(ByteBuf buf)
     {
-        uuid = readString( buf );
+        // Waterfall (1.16: 20w12a+ UUID is now a four-int array) start
+        if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            uuid = readUUIDIntArray( buf ).toString();
+        } else {
+            uuid = readString( buf );
+        }
+        // Waterfall end
         username = readString( buf );
     }
 
     @Override
     public void write(ByteBuf buf)
     {
-        writeString( uuid, buf );
+        // Waterfall (1.16: 20w12a+ UUID is now a four-int array) start
+        if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            writeUUIDIntArray( UUID.fromString( uuid ), buf);
+        } else {
+            writeString( uuid, buf );
+        }
+        // Waterfall end
         writeString( username, buf );
     }
 
@@ -37,4 +55,18 @@ public class LoginSuccess extends DefinedPacket
     {
         handler.handle( this );
     }
+
+    // Waterfall (1.16: 20w12a+ forward protocol version) start
+    @Override
+    public void read(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion){
+        this.protocolVersion = protocolVersion;
+        read(buf);
+    }
+
+    @Override
+    public void write(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion){
+        this.protocolVersion = protocolVersion;
+        write(buf);
+    }
+    // Waterfall end
 }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Respawn.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Respawn.java
index 1f544f26..e6e1ecd2 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/Respawn.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/Respawn.java
@@ -21,11 +21,25 @@ public class Respawn extends DefinedPacket
     private short difficulty;
     private short gameMode;
     private String levelType;
+    private boolean keepAllPlayerData; // Waterfall (1.16: 20w18a+)
+    private boolean isDebug; // Waterfall (1.16: 20w20a+)
+    private boolean isFlat; // Waterfall (1.16: 20w20a+)
+    private String dimensionCodecName; // Waterfall (1.16: 20w21a+)
+    private String world; // Waterfall (1.16: 20w22a+)
 
     @Override
     public void read(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion)
     {
-        dimension = buf.readInt();
+        // Waterfall (1.16: 20w21a+) start:
+        if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            dimensionCodecName = readString( buf );
+            world = readString( buf ); // Waterfall (1.16: 20w22a+)
+        } else
+        {
+            dimension = buf.readInt();
+        }
+        // Waterfall end
         if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_15 )
         {
             seed = buf.readLong();
@@ -35,13 +49,32 @@ public class Respawn extends DefinedPacket
             difficulty = buf.readUnsignedByte();
         }
         gameMode = buf.readUnsignedByte();
-        levelType = readString( buf );
+        // Waterfall (1.16: 20w20+) start:
+        if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            isDebug = buf.readBoolean();
+            isFlat = buf.readBoolean();
+            keepAllPlayerData = buf.readBoolean();
+        } else
+        {
+            levelType = readString( buf );
+        }
+        // Waterfall end
     }
 
     @Override
     public void write(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion)
     {
-        buf.writeInt( dimension );
+        // Waterfall (1.16: 20w21a+) start:
+        if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+           writeString( dimensionCodecName, buf );
+           writeString( world, buf ); // Waterfall (1.16: 20w22a+)
+        } else
+        {
+            buf.writeInt( dimension );
+        }
+        // Waterfall end
         if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_15 )
         {
             buf.writeLong( seed );
@@ -51,7 +84,17 @@ public class Respawn extends DefinedPacket
             buf.writeByte( difficulty );
         }
         buf.writeByte( gameMode );
-        writeString( levelType, buf );
+        // Waterfall (1.16: 20w18a+) start:
+        if( protocolVersion >= ProtocolConstants.MINECRAFT_1_16 )
+        {
+            buf.writeBoolean( isDebug );
+            buf.writeBoolean( isFlat );
+            buf.writeBoolean( keepAllPlayerData );
+        } else
+        {
+            writeString( levelType, buf );
+        }
+        // Waterfall end
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/PacketConstants.java b/proxy/src/main/java/net/md_5/bungee/PacketConstants.java
index 000c450e..5f372766 100644
--- a/proxy/src/main/java/net/md_5/bungee/PacketConstants.java
+++ b/proxy/src/main/java/net/md_5/bungee/PacketConstants.java
@@ -7,8 +7,8 @@ import net.md_5.bungee.protocol.packet.Respawn;
 public class PacketConstants
 {
 
-    public static final Respawn DIM1_SWITCH = new Respawn( (byte) 1, 0, (byte) 0, (byte) 0, "default" );
-    public static final Respawn DIM2_SWITCH = new Respawn( (byte) -1, 0, (byte) 0, (byte) 0, "default" );
+    public static final Respawn DIM1_SWITCH = new Respawn( (byte) 1, 0, (byte) 0, (byte) 0, "default", false, /* Waterfall (1.16: 20w18a+) */ false, false, /* Waterfall 1.16 (20w20a+)*/ "minecraft:overworld", /* Waterfall 1.16 (20w21a+)*/ "bungeecord:dimension_switch1" /* Waterfall 1.16 (20w22a+)*/ );
+    public static final Respawn DIM2_SWITCH = new Respawn( (byte) -1, 0, (byte) 0, (byte) 0, "default", false, /* Waterfall (1.16: 20w18a+) */ false, false, /* Waterfall 1.16 (20w20a+)*/ "minecraft:overworld", /* Waterfall 1.16 (20w21a+)*/ "bungeecord:dimension_switch1" /* Waterfall 1.16 (20w22a+)*/ );
     public static final ClientStatus CLIENT_LOGIN = new ClientStatus( (byte) 0 );
     public static final PluginMessage FORGE_MOD_REQUEST = new PluginMessage( "FML", new byte[]
     {
diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index 35a19224..1e8dfe09 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -246,7 +246,7 @@ public class ServerConnector extends PacketHandler
 
             // Set tab list size, TODO: what shall we do about packet mutability
             Login modLogin = new Login( login.getEntityId(), login.getGameMode(), (byte) login.getDimension(), login.getSeed(), login.getDifficulty(),
-                    (byte) user.getPendingConnection().getListener().getTabListSize(), login.getLevelType(), login.getViewDistance(), login.isReducedDebugInfo(), login.isNormalRespawn() );
+                    (byte) user.getPendingConnection().getListener().getTabListSize(), login.getLevelType(), login.getViewDistance(), login.isReducedDebugInfo(), login.isNormalRespawn(), login.isDebug(), login.isFlat(), /* Waterfall 1.16 (20w20a+)*/ login.getDimensionCodecName(), login.getCodecTag(), /* Waterfall 1.16 (20w21a+)*/ login.getWorlds(), login.getWorld() /* Waterfall 1.16 (20w22a+)*/ );
 
             user.unsafe().sendPacket( modLogin );
 
@@ -296,7 +296,7 @@ public class ServerConnector extends PacketHandler
             user.setDimensionChange( true );
             if ( !user.isDisableEntityMetadataRewrite() && login.getDimension() == user.getDimension() ) // Waterfall - defer
             {
-                user.unsafe().sendPacket( new Respawn( ( login.getDimension() >= 0 ? -1 : 0 ), login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType() ) );
+                user.unsafe().sendPacket( new Respawn( ( login.getDimension() >= 0 ? -1 : 0 ), login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType(), false,  /* Waterfall (1.16: 20w18a+) */ login.isDebug(), login.isFlat(), /* Waterfall 1.16 (20w20a+)*/ login.getDimensionCodecName(), /* Waterfall 1.16 (20w21a+)*/ login.getWorld() /* Waterfall 1.16 (20w22a+)*/ ) );
             }
 
             user.setServerEntityId( login.getEntityId() );
@@ -308,20 +308,20 @@ public class ServerConnector extends PacketHandler
 
                 // Only send if we are not in the same dimension
                 if (user.getDimension() != login.getDimension()) {
-                    user.unsafe().sendPacket( new Respawn(user.getDimension() == 0 ? -1 : 0, login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType()));
+                    user.unsafe().sendPacket( new Respawn(user.getDimension() == 0 ? -1 : 0, login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType(), false,  /* Waterfall (1.16: 20w18a+) */ login.isDebug(), login.isFlat(),  /* Waterfall 1.16 (20w20a+)*/ login.getDimensionCodecName(), /* Waterfall 1.16 (20w21a+)*/ login.getWorld() /* Waterfall 1.16 (20w22a+)*/  ));
                 }
 
                 Login modLogin = new Login( login.getEntityId(),login.getGameMode(), login.getDimension(), login.getSeed(),
-                        login.getDifficulty(), login.getMaxPlayers(), login.getLevelType(), login.getViewDistance(), login.isReducedDebugInfo(), login.isNormalRespawn() );
+                        login.getDifficulty(), login.getMaxPlayers(), login.getLevelType(), login.getViewDistance(), login.isReducedDebugInfo(), login.isNormalRespawn(), login.isDebug(), login.isFlat(), /* Waterfall 1.16 (20w20a+)*/ login.getDimensionCodecName(), login.getCodecTag(), /* Waterfall 1.16 (20w21a+)*/ login.getWorlds(), login.getWorld() /* Waterfall 1.16 (20w22a+)*/ );
                 user.unsafe().sendPacket(modLogin);
 
                 // Only send if we're in the same dimension
                 if (user.getDimension() == login.getDimension()) {
-                    user.unsafe().sendPacket( new Respawn(user.getDimension() == 0 ? -1 : 0, login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType()));
+                    user.unsafe().sendPacket( new Respawn(user.getDimension() == 0 ? -1 : 0, login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType(), false,  /* Waterfall (1.16: 20w18a+) */ login.isDebug(), login.isFlat(),  /* Waterfall 1.16 (20w20a+)*/ login.getDimensionCodecName(), /* Waterfall 1.16 (20w21a+)*/ login.getWorld() /* Waterfall 1.16 (20w22a+)*/ ));
                 }
             }
             // Waterfall end
-            user.unsafe().sendPacket( new Respawn( login.getDimension(), login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType() ) );
+            user.unsafe().sendPacket( new Respawn( login.getDimension(), login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getLevelType(), false,  /* Waterfall (1.16: 20w18a+) */ login.isDebug(), login.isFlat(), /* Waterfall 1.16 (20w20a+)*/ login.getDimensionCodecName(), /* Waterfall 1.16 (20w21a+)*/ login.getWorld() /* Waterfall 1.16 (20w22a+)*/ ) );
             if ( user.getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_14 )
             {
                 user.unsafe().sendPacket( new ViewDistance( login.getViewDistance() ) );
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 87cff4ff..3fb9bc16 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -465,7 +465,7 @@ public final class UserConnection implements ProxiedPlayer
 
     private void sendMessage(ChatMessageType position, String message)
     {
-        unsafe().sendPacket( new Chat( message, (byte) position.ordinal() ) );
+        unsafe().sendPacket( new Chat( message, (byte) position.ordinal(), new UUID( 0, 0 ) /* Waterfall 1.16 (20w21a+)*/ ) );
     }
 
     @Override
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 37ea35e3..98c0c2d7 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -536,7 +536,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                             userCon.setCompressionThreshold( BungeeCord.getInstance().config.getCompressionThreshold() );
                             userCon.init();
 
-                            unsafe.sendPacket( new LoginSuccess( getUniqueId().toString(), getName() ) ); // With dashes in between
+                            unsafe.sendPacket( new LoginSuccess( getUniqueId().toString(), getName(), getVersion() /* Waterfall (1.16: 20w12a+) */ ) ); // With dashes in between
                             ch.setProtocol( Protocol.GAME );
 
                             ch.getHandle().pipeline().get( HandlerBoss.class ).setHandler( new UpstreamBridge( bungee, userCon ) );
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
index bc1de55a..ef2a66fa 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
@@ -67,6 +67,8 @@ public abstract class EntityMap
             case ProtocolConstants.MINECRAFT_1_15_1:
             case ProtocolConstants.MINECRAFT_1_15_2:
                 return EntityMap_1_15.INSTANCE;
+            case ProtocolConstants.MINECRAFT_1_16:
+                return EntityMap_1_16.INSTANCE;
         }
         throw new RuntimeException( "Version " + version + " has no entity map" );
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
new file mode 100644
index 00000000..e87d2632
--- /dev/null
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap_1_16.java
@@ -0,0 +1,194 @@
+// Travertine start
+package net.md_5.bungee.entitymap;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import io.netty.buffer.ByteBuf;
+import net.md_5.bungee.BungeeCord;
+import net.md_5.bungee.UserConnection;
+import net.md_5.bungee.api.connection.ProxiedPlayer;
+import net.md_5.bungee.protocol.DefinedPacket;
+import net.md_5.bungee.protocol.ProtocolConstants;
+
+import java.util.UUID;
+
+class EntityMap_1_16 extends EntityMap
+{
+
+    static final EntityMap_1_16 INSTANCE = new EntityMap_1_16();
+
+    EntityMap_1_16()
+    {
+        addRewrite( 0x00, ProtocolConstants.Direction.TO_CLIENT, true ); // Spawn Object : PacketPlayOutSpawnEntity
+        addRewrite( 0x01, ProtocolConstants.Direction.TO_CLIENT, true ); // Spawn Experience Orb : PacketPlayOutSpawnEntityExperienceOrb
+        addRewrite( 0x03, ProtocolConstants.Direction.TO_CLIENT, true ); // Spawn Mob : PacketPlayOutSpawnEntityLiving
+        addRewrite( 0x04, ProtocolConstants.Direction.TO_CLIENT, true ); // Spawn Painting : PacketPlayOutSpawnEntityPainting
+        addRewrite( 0x05, ProtocolConstants.Direction.TO_CLIENT, true ); // Spawn Player : PacketPlayOutNamedEntitySpawn
+        addRewrite( 0x06, ProtocolConstants.Direction.TO_CLIENT, true ); // Animation : PacketPlayOutAnimation
+        addRewrite( 0x09, ProtocolConstants.Direction.TO_CLIENT, true ); // Block Break Animation : PacketPlayOutBlockBreakAnimation
+        addRewrite( 0x1C, ProtocolConstants.Direction.TO_CLIENT, false ); // Entity Status : PacketPlayOutEntityStatus
+        addRewrite( 0x29, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Relative Move : PacketPlayOutRelEntityMove
+        addRewrite( 0x2A, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Look and Relative Move : PacketPlayOutRelEntityMoveLook
+        addRewrite( 0x2B, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Look : PacketPlayOutEntityLook
+        addRewrite( 0x2C, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity : PacketPlayOutEntity
+        addRewrite( 0x39, ProtocolConstants.Direction.TO_CLIENT, true ); // Remove Entity Effect : PacketPlayOutRemoveEntityEffect
+        addRewrite( 0x3C, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Head Look : PacketPlayOutEntityHeadRotation
+        addRewrite( 0x3F, ProtocolConstants.Direction.TO_CLIENT, true ); // Camera : PacketPlayOutCamera
+        /* Waterfall (1.16: 20w12a+ 0x44 -> 0x45) */ addRewrite( 0x45, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Metadata : PacketPlayOutEntityMetadata
+        /* Waterfall (1.16: 20w12a+ 0x45 -> 0x46) */ addRewrite( 0x46, ProtocolConstants.Direction.TO_CLIENT, false ); // Attach Entity : PacketPlayOutAttachEntity
+        /* Waterfall (1.16: 20w12a+ 0x46 -> 0x47) */ addRewrite( 0x47, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Velocity : PacketPlayOutEntityVelocity
+        /* Waterfall (1.16: 20w12a+ 0x47 -> 0x48) */ addRewrite( 0x48, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Equipment : PacketPlayOutEntityEquipment
+        /* Waterfall (1.16: 20w12a+ 0x4B -> 0x4C) */ addRewrite( 0x4C, ProtocolConstants.Direction.TO_CLIENT, true ); // Set Passengers : PacketPlayOutMount
+        addRewrite( 0x56, ProtocolConstants.Direction.TO_CLIENT, true ); // Collect Item : PacketPlayOutCollect
+        addRewrite( 0x57, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Teleport : PacketPlayOutEntityTeleport
+        addRewrite( 0x59, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Properties : PacketPlayOutUpdateAttributes
+        addRewrite( 0x5A, ProtocolConstants.Direction.TO_CLIENT, true ); // Entity Effect : PacketPlayOutEntityEffect
+
+        addRewrite( 0x0E, ProtocolConstants.Direction.TO_SERVER, true ); // Use Entity : PacketPlayInUseEntity
+        /* Waterfall (1.16: 20w16a+ 0x1B -> 0x1C) */  addRewrite( 0x1C, ProtocolConstants.Direction.TO_SERVER, true ); // Entity Action : PacketPlayInEntityAction
+    }
+
+    @Override
+    @SuppressFBWarnings("DLS_DEAD_LOCAL_STORE")
+    public void rewriteClientbound(ByteBuf packet, int oldId, int newId, int protocolVersion)
+    {
+        super.rewriteClientbound( packet, oldId, newId );
+
+        // Special cases
+        int readerIndex = packet.readerIndex();
+        int packetId = DefinedPacket.readVarInt( packet );
+        int packetIdLength = packet.readerIndex() - readerIndex;
+        int jumpIndex = packet.readerIndex();
+        switch ( packetId )
+        {
+            /* Waterfall (1.16: 20w12a+ 0x45 -> 0x46) */ case 0x46 /* Attach Entity : PacketPlayOutAttachEntity */:
+            rewriteInt( packet, oldId, newId, readerIndex + packetIdLength + 4 );
+            break;
+            case 0x56 /* Collect Item : PacketPlayOutCollect */:
+                DefinedPacket.readVarInt( packet );
+                rewriteVarInt( packet, oldId, newId, packet.readerIndex() );
+                break;
+            case 0x4C /* Set Passengers : PacketPlayOutMount */:
+                DefinedPacket.readVarInt( packet );
+                jumpIndex = packet.readerIndex();
+                // Fall through on purpose to int array of IDs
+            case 0x38 /* Destroy Entities : PacketPlayOutEntityDestroy */:
+                int count = DefinedPacket.readVarInt( packet );
+                int[] ids = new int[ count ];
+                for ( int i = 0; i < count; i++ )
+                {
+                    ids[i] = DefinedPacket.readVarInt( packet );
+                }
+                packet.readerIndex( jumpIndex );
+                packet.writerIndex( jumpIndex );
+                DefinedPacket.writeVarInt( count, packet );
+                for ( int id : ids )
+                {
+                    if ( id == oldId )
+                    {
+                        id = newId;
+                    } else if ( id == newId )
+                    {
+                        id = oldId;
+                    }
+                    DefinedPacket.writeVarInt( id, packet );
+                }
+                break;
+            case 0x00 /* Spawn Object : PacketPlayOutSpawnEntity */:
+                DefinedPacket.readVarInt( packet );
+                DefinedPacket.readUUID( packet );
+                int type = DefinedPacket.readVarInt( packet );
+
+                if ( type == 2 || type == 102 || type == 72 ) // arrow, fishing_bobber or spectral_arrow
+                {
+                    if ( type == 2 || type == 72 ) // arrow or spectral_arrow
+                    {
+                        oldId = oldId + 1;
+                        newId = newId + 1;
+                    }
+
+                    packet.skipBytes( 26 ); // double, double, double, byte, byte
+                    int position = packet.readerIndex();
+                    int readId = packet.readInt();
+                    if ( readId == oldId )
+                    {
+                        packet.setInt( position, newId );
+                    } else if ( readId == newId )
+                    {
+                        packet.setInt( position, oldId );
+                    }
+                }
+                break;
+            case 0x05 /* Spawn Player : PacketPlayOutNamedEntitySpawn */:
+                DefinedPacket.readVarInt( packet ); // Entity ID
+                int idLength = packet.readerIndex() - readerIndex - packetIdLength;
+                UUID uuid = DefinedPacket.readUUID( packet );
+                ProxiedPlayer player;
+                if ( ( player = BungeeCord.getInstance().getPlayerByOfflineUUID( uuid ) ) != null )
+                {
+                    int previous = packet.writerIndex();
+                    packet.readerIndex( readerIndex );
+                    packet.writerIndex( readerIndex + packetIdLength + idLength );
+                    DefinedPacket.writeUUID( player.getUniqueId(), packet );
+                    packet.writerIndex( previous );
+                }
+                break;
+            case 0x33 /* Combat Event : PacketPlayOutCombatEvent */:
+                int event = packet.readUnsignedByte();
+                if ( event == 1 /* End Combat*/ )
+                {
+                    DefinedPacket.readVarInt( packet );
+                    rewriteInt( packet, oldId, newId, packet.readerIndex() );
+                } else if ( event == 2 /* Entity Dead */ )
+                {
+                    int position = packet.readerIndex();
+                    rewriteVarInt( packet, oldId, newId, packet.readerIndex() );
+                    packet.readerIndex( position );
+                    DefinedPacket.readVarInt( packet );
+                    rewriteInt( packet, oldId, newId, packet.readerIndex() );
+                }
+                break;
+            /* Waterfall (1.16: 20w12a+ 0x44 -> 0x45) */ case 0x45 /* EntityMetadata : PacketPlayOutEntityMetadata */:
+            DefinedPacket.readVarInt( packet ); // Entity ID
+            rewriteMetaVarInt( packet, oldId + 1, newId + 1, 7, protocolVersion ); // fishing hook
+            /*
+            * Waterfall (1.16: 20w12a+)
+            * metaVarInt was shifted to 8 from 16 and 16 omitted, previously:
+            * rewriteMetaVarInt( packet, oldId, newId, 8, protocolVersion ); // fireworks (et al)
+            * */
+            rewriteMetaVarInt( packet, oldId, newId, 8, protocolVersion ); // guardian beam
+            break;
+            case 0x51 /* Entity Sound Effect : PacketPlayOutEntitySound */:
+                DefinedPacket.readVarInt( packet );
+                DefinedPacket.readVarInt( packet );
+                rewriteVarInt( packet, oldId, newId, packet.readerIndex() );
+                break;
+        }
+        packet.readerIndex( readerIndex );
+    }
+
+    @Override
+    public void rewriteServerbound(ByteBuf packet, int oldId, int newId)
+    {
+        super.rewriteServerbound( packet, oldId, newId );
+        // Special cases
+        int readerIndex = packet.readerIndex();
+        int packetId = DefinedPacket.readVarInt( packet );
+        int packetIdLength = packet.readerIndex() - readerIndex;
+
+        if ( packetId == 0x2C /* Waterfall (1.16: 20w16a+ 0x2B -> 0x2C) Spectate : PacketPlayInSpectate */ && !BungeeCord.getInstance().getConfig().isIpForward() )
+        {
+            UUID uuid = DefinedPacket.readUUID( packet );
+            ProxiedPlayer player;
+            if ( ( player = BungeeCord.getInstance().getPlayer( uuid ) ) != null )
+            {
+                int previous = packet.writerIndex();
+                packet.readerIndex( readerIndex );
+                packet.writerIndex( readerIndex + packetIdLength );
+                DefinedPacket.writeUUID( ( (UserConnection) player ).getPendingConnection().getOfflineId(), packet );
+                packet.writerIndex( previous );
+            }
+        }
+        packet.readerIndex( readerIndex );
+    }
+}
+// Travertine end
\ No newline at end of file
-- 
2.24.0

