From e668f69d4329b3fbdf1d4a088c42edb48b898e57 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sat, 9 May 2020 21:40:23 -0400
Subject: [PATCH] Cap tab complete packets to servers max


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/packet/TabCompleteRequest.java b/protocol/src/main/java/net/md_5/bungee/protocol/packet/TabCompleteRequest.java
index d6865ae2..3e4ea192 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/packet/TabCompleteRequest.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/packet/TabCompleteRequest.java
@@ -41,7 +41,7 @@ public class TabCompleteRequest extends DefinedPacket
         {
             transactionId = readVarInt( buf );
         }
-        cursor = readString( buf );
+        cursor = readString( buf , 32500); // Waterfall
 
         if ( protocolVersion < ProtocolConstants.MINECRAFT_1_13 )
         {
-- 
2.28.0

