From 2cd21b7b20cde396fa6d8713116f055dce6b72f4 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Fri, 9 Jul 2021 13:20:36 +0100
Subject: [PATCH] Don't bother locking to fetch a v4 UUID from the offline
 UUIDs map


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 15fb4ad2..e0fcd8fb 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -634,6 +634,11 @@ public class BungeeCord extends ProxyServer
 
     public UserConnection getPlayerByOfflineUUID(UUID name)
     {
+        // Waterfall start - Skip lock if we get a v4 UUID, we're not gonna be in here
+        if (name != null && name.version() == 4) {
+            return null;
+        }
+        // Waterfall end
         connectionLock.readLock().lock();
         try
         {
-- 
2.38.1

