From d3bf43998d56d5050eb34fded366e0fdb3ae4f4e Mon Sep 17 00:00:00 2001
From: "Five (Xer)" <admin@xernium.com>
Date: Thu, 10 Sep 2020 17:42:01 +0200
Subject: [PATCH] Waterfall-1.16.3-Patch


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index aa1b770c..f33bb5c4 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -31,6 +31,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_16 = 735;
     public static final int MINECRAFT_1_16_1 = 736;
     public static final int MINECRAFT_1_16_2 = 751;
+    public static final int MINECRAFT_1_16_3 = 753;
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.8.x",
             "1.9.x",
@@ -67,7 +68,8 @@ public class ProtocolConstants
             ProtocolConstants.MINECRAFT_1_15_2,
             ProtocolConstants.MINECRAFT_1_16,
             ProtocolConstants.MINECRAFT_1_16_1,
-            ProtocolConstants.MINECRAFT_1_16_2
+            ProtocolConstants.MINECRAFT_1_16_2,
+            ProtocolConstants.MINECRAFT_1_16_3
     );
 
     public static final boolean isBeforeOrEq(int before, int other)
-- 
2.24.0

