From 67d6af783dc584fccb4a7bef8358c71e3f444b95 Mon Sep 17 00:00:00 2001
From: Gerrygames <gecam59@gmail.com>
Date: Mon, 12 Oct 2020 11:11:26 +0200
Subject: [PATCH] Fix infinite relay message accumulation


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 4d7b1b23..78aff245 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -269,7 +269,7 @@ public class UpstreamBridge extends PacketHandler
         }
 
         // TODO: Unregister as well?
-        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) )
+        if ( PluginMessage.SHOULD_RELAY.apply( pluginMessage ) && !con.getPendingConnection().getRelayMessages().contains( pluginMessage ) ) // Waterfall - add contains check
         {
             con.getPendingConnection().getRelayMessages().add( pluginMessage );
         }
-- 
2.20.1

