From 4ae1a8331d069a8c694700006553126c1a9fb287 Mon Sep 17 00:00:00 2001
From: LemonCaramel <admin@caramel.moe>
Date: Fri, 17 Sep 2021 03:59:41 +0900
Subject: [PATCH] Preliminary 1.18 Snapshot support


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index bb37ba04..4dea1d3f 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -36,7 +36,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_16_4 = 754;
     public static final int MINECRAFT_1_17 = 755;
     public static final int MINECRAFT_1_17_1 = 756;
-    public static final int MINECRAFT_1_18 = 1073741868;
+    public static final int MINECRAFT_1_18 = (1 << 30) | 47;
     public static final List<String> SUPPORTED_VERSIONS;
     public static final List<Integer> SUPPORTED_VERSION_IDS;
 
@@ -52,7 +52,8 @@ public class ProtocolConstants
                 "1.14.x",
                 "1.15.x",
                 "1.16.x",
-                "1.17.x"
+                "1.17.x",
+                "1.18"
         );
         ImmutableList.Builder<Integer> supportedVersionIds = ImmutableList.<Integer>builder().add(
                 ProtocolConstants.MINECRAFT_1_8,
@@ -83,7 +84,8 @@ public class ProtocolConstants
                 ProtocolConstants.MINECRAFT_1_16_3,
                 ProtocolConstants.MINECRAFT_1_16_4,
                 ProtocolConstants.MINECRAFT_1_17,
-                ProtocolConstants.MINECRAFT_1_17_1
+                ProtocolConstants.MINECRAFT_1_17_1,
+                ProtocolConstants.MINECRAFT_1_18
         );
 
         if ( SNAPSHOT_SUPPORT )
-- 
2.28.0.windows.1

