From 5b9b70f0fd6a5bb5ea216fde5b24924d74e018a6 Mon Sep 17 00:00:00 2001
From: Lorenzo <69311874+Lorenzo0111@users.noreply.github.com>
Date: Fri, 9 Jul 2021 17:38:27 +0200
Subject: [PATCH] Added messages file extractor


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 07d74c67..76a4cd64 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -21,12 +21,11 @@ import io.netty.channel.ChannelFutureListener;
 import io.netty.channel.ChannelOption;
 import io.netty.channel.EventLoopGroup;
 import io.netty.util.ResourceLeakDetector;
-import java.io.File;
-import java.io.FileReader;
-import java.io.IOException;
-import java.io.PrintStream;
+
+import java.io.*;
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
+import java.nio.file.Files;
 import java.text.MessageFormat;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -197,6 +196,20 @@ public class BungeeCord extends ProxyServer
         {
             baseBundle = ResourceBundle.getBundle( "messages", Locale.ENGLISH );
         }
+
+        // Waterfall start
+        File file = new File( "messages.properties" );
+        if (!file.exists()) {
+            try (InputStream in = this.getClass().getClassLoader().getResourceAsStream( "messages.properties" )) {
+                if (in != null) {
+                    Files.copy(in, file.toPath());
+                }
+            } catch (IOException e) {
+                getLogger().log( Level.SEVERE, "Could not load custom messages.properties", e );
+            }
+        }
+        // Waterfall end
+
         reloadMessages();
 
         // This is a workaround for quite possibly the weirdest bug I have ever encountered in my life!
-- 
2.32.0.windows.2

