From 4f62e9b808995e99b6c081b595f0fd937566c5ff Mon Sep 17 00:00:00 2001
From: Nassim Jahnke <jahnke.nassim@gmail.com>
Date: Tue, 30 Nov 2021 10:24:30 +0100
Subject: [PATCH] 1.18


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 8422f88b..f1276c09 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -36,7 +36,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_16_4 = 754;
     public static final int MINECRAFT_1_17 = 755;
     public static final int MINECRAFT_1_17_1 = 756;
-    public static final int MINECRAFT_1_18 = 1073741883;
+    public static final int MINECRAFT_1_18 = 757; // Waterfall
     public static final List<String> SUPPORTED_VERSIONS;
     public static final List<Integer> SUPPORTED_VERSION_IDS;
 
-- 
2.34.1.windows.1

