From d3343b033840d723bfc103fca04c04cb2975c412 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Fri, 21 Apr 2023 15:32:33 +0100
Subject: [PATCH] reduce log spam from clients registeirng too many channels


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index a9358a51..5e74d0e2 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -162,7 +162,17 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     @Override
     public void handle(PluginMessage pluginMessage) throws Exception
     {
-        this.relayMessage( pluginMessage );
+        // Waterfall start
+        try {
+            this.relayMessage(pluginMessage);
+        } catch (IllegalStateException | IllegalArgumentException ex) {
+            if (net.md_5.bungee.protocol.MinecraftDecoder.DEBUG) {
+                throw ex;
+            } else {
+                throw new QuietException(ex.getMessage());
+            }
+        }
+        // Waterfall end
     }
 
     @Override
-- 
2.43.0

