From 44c080b7d0505a30ea7e7f501fe1298b6a89e2db Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 17 Jun 2025 15:04:22 +0100
Subject: [PATCH] Temp release support


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 1c1a09bc..9c824dcf 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -119,7 +119,7 @@ public class ProtocolConstants
                 ProtocolConstants.MINECRAFT_1_21_6
         );
 
-        if ( SNAPSHOT_SUPPORT )
+        if ( true || SNAPSHOT_SUPPORT )
         {
             // supportedVersions.add( "1.21.x" );
             // supportedVersionIds.add( ProtocolConstants.MINECRAFT_1_21_6 );
-- 
2.50.0

