From 70041a58b886a27ed67c4318b8ce8e2445908780 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 17 Jun 2025 15:04:22 +0100
Subject: [PATCH] Temp release support


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 8760f52d..edf78be4 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -50,7 +50,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_21_2 = 768;
     public static final int MINECRAFT_1_21_4 = 769;
     public static final int MINECRAFT_1_21_5 = 770;
-    public static final int MINECRAFT_1_21_6 = 1073742078;
+    public static final int MINECRAFT_1_21_6 = 771;
     public static final List<String> SUPPORTED_VERSIONS;
     public static final List<Integer> SUPPORTED_VERSION_IDS;
 
@@ -118,7 +118,7 @@ public class ProtocolConstants
                 ProtocolConstants.MINECRAFT_1_21_5
         );
 
-        if ( SNAPSHOT_SUPPORT )
+        if ( true || SNAPSHOT_SUPPORT )
         {
             // supportedVersions.add( "1.21.x" );
             supportedVersionIds.add( ProtocolConstants.MINECRAFT_1_21_6 );
-- 
2.49.0

