From 658a0d875f9b5dd28a442ae83e42db2cb9456da6 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Thu, 21 Sep 2023 11:02:43 +0100
Subject: [PATCH] fixup! Fix potion race condition on Forge 1.8.9


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 485af04f..76acae0e 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -745,7 +745,7 @@ public class DownstreamBridge extends PacketHandler
 
     // Waterfall start
     @Override
-    public void handle(EntityEffect entityEffect) throws Exception
+    public void handle(net.md_5.bungee.protocol.packet.EntityEffect entityEffect) throws Exception
     {
         if (con.isDisableEntityMetadataRewrite()) return; // Waterfall
         // Don't send any potions when switching between servers (which involves a handshake), which can trigger a race
@@ -757,7 +757,7 @@ public class DownstreamBridge extends PacketHandler
     }
 
     @Override
-    public void handle(EntityRemoveEffect removeEffect) throws Exception
+    public void handle(net.md_5.bungee.protocol.packet.EntityRemoveEffect removeEffect) throws Exception
     {
         if (con.isDisableEntityMetadataRewrite()) return; // Waterfall
         con.getPotions().remove(rewriteEntityId(removeEffect.getEntityId()), removeEffect.getEffectId());
-- 
2.42.0

