From ff461b8c47aadc617c4c7f322e44db80473906db Mon Sep 17 00:00:00 2001
From: Techcable <Techcable@techcable.net>
Date: Mon, 6 Jun 2016 13:49:46 -0600
Subject: [PATCH] Metrics


diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/Metrics.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/Metrics.java
index ae5a2a9..4b70e8c 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/Metrics.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/Metrics.java
@@ -73,7 +73,7 @@ public class Metrics extends TimerTask
         }
 
         // Create the url
-        URL url = new URL( BASE_URL + String.format( REPORT_URL, encode( "Waterfall" ) ) );
+        URL url = new URL( BASE_URL + String.format( REPORT_URL, encode( "Travertine" ) ) );
 
         // Connect to the website
         URLConnection connection;
-- 
2.8.3

