From e8c93c149c23d9d6d57acee204780c952a3394bf Mon Sep 17 00:00:00 2001
From: "Five (Xer)" <admin@xernium.com>
Date: Thu, 10 Sep 2020 17:59:48 +0200
Subject: [PATCH] Travertine-1.16.3-Patch


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 6f72983f..659382c0 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -33,6 +33,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_16 = 735;
     public static final int MINECRAFT_1_16_1 = 736;
     public static final int MINECRAFT_1_16_2 = 751;
+    public static final int MINECRAFT_1_16_3 = 753;
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.7.x",
             "1.8.x",
@@ -72,7 +73,8 @@ public class ProtocolConstants
             ProtocolConstants.MINECRAFT_1_15_2,
             ProtocolConstants.MINECRAFT_1_16,
             ProtocolConstants.MINECRAFT_1_16_1,
-            ProtocolConstants.MINECRAFT_1_16_2
+            ProtocolConstants.MINECRAFT_1_16_2,
+            ProtocolConstants.MINECRAFT_1_16_3
     );
 
     public static final boolean isBeforeOrEq(int before, int other)
-- 
2.24.0

