From 573492c8d5a5a6513a7a2be8e84dcbb3d5227b3f Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Tue, 21 Jan 2020 17:44:04 +0000
Subject: [PATCH] TEMP 1.15.2


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index be966c44..68327078 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -29,6 +29,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_14_4 = 498;
     public static final int MINECRAFT_1_15 = 573;
     public static final int MINECRAFT_1_15_1 = 575;
+    public static final int MINECRAFT_1_15_2 = 578; // Waterfall
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.7.x",
             "1.8.x",
@@ -63,7 +64,8 @@ public class ProtocolConstants
             ProtocolConstants.MINECRAFT_1_14_3,
             ProtocolConstants.MINECRAFT_1_14_4,
             ProtocolConstants.MINECRAFT_1_15,
-            ProtocolConstants.MINECRAFT_1_15_1
+            ProtocolConstants.MINECRAFT_1_15_1,
+            ProtocolConstants.MINECRAFT_1_15_2 // Waterfall
     );
 
     public static final boolean isBeforeOrEq(int before, int other)
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
index 67985446..91832f7e 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
@@ -68,6 +68,7 @@ public abstract class EntityMap
                 return EntityMap_1_14.INSTANCE;
             case ProtocolConstants.MINECRAFT_1_15:
             case ProtocolConstants.MINECRAFT_1_15_1:
+            case ProtocolConstants.MINECRAFT_1_15_2: // Waterfall
                 return EntityMap_1_15.INSTANCE;
         }
         throw new RuntimeException( "Version " + version + " has no entity map" );
-- 
2.25.0

