From b7a9aef9f7b615eb356879eb367e42baf9426a20 Mon Sep 17 00:00:00 2001
From: Troy Frew <fuzzy_bot@arenaga.me>
Date: Tue, 15 Nov 2016 10:31:04 -0500
Subject: [PATCH] 1.7.x Protocol Patch


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 2aa6885..2ff8a6a 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -274,6 +274,8 @@ public enum Protocol
         private final TIntObjectMap<List<Integer>> linkedProtocols = new TIntObjectHashMap<>();
         {
             linkedProtocols.put( ProtocolConstants.MINECRAFT_1_8, Arrays.asList(
+                    ProtocolConstants.MINECRAFT_1_7_2,
+                    ProtocolConstants.MINECRAFT_1_7_6,
                     ProtocolConstants.MINECRAFT_1_9
             ) );
             linkedProtocols.put( ProtocolConstants.MINECRAFT_1_9, Arrays.asList(
@@ -312,6 +314,9 @@ public enum Protocol
             }
             if ( !hasPacket(id, supportsForge) )
             {
+                if ( ProtocolConstants.isBeforeOrEq( version, ProtocolConstants.MINECRAFT_1_7_6 ) ) {
+                    return null;
+                }
                 throw new BadPacketException( "Packet with id " + id + " outside of range " );
             }
 
-- 
2.7.4

