From efe1b091836279bb375f6770fb8d98c4e5c1c4ed Mon Sep 17 00:00:00 2001
From: Juan Cruz Linsalata <LinsaFTW@users.noreply.github.com>
Date: Thu, 3 Sep 2020 19:36:44 -0300
Subject: [PATCH] FlameCord Defaulted timeout from 30000 to 5000


diff --git a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
index 85d2197a..a984370e 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
@@ -29,13 +29,12 @@ import net.md_5.bungee.util.CaseInsensitiveSet;
  * Core configuration for the proxy.
  */
 @Getter
-public abstract class Configuration implements ProxyConfig
-{
-
+public abstract class Configuration implements ProxyConfig {
+    // FlameCord - Default from 30000 to 5000
     /**
      * Time before users are disconnected due to no network activity.
      */
-    private int timeout = 30000;
+    private int timeout = 5000;
     /**
      * UUID used for metrics.
      */
--
2.27.0.windows.1

