From 2c2ada407ef6e402318ccd308384210059266f92 Mon Sep 17 00:00:00 2001
From: Shane Freeder <theboyetronic@gmail.com>
Date: Fri, 20 Jul 2018 03:54:51 +0100
Subject: [PATCH] Add temp 1.13 release support

To be dropped when upstream releases final support for 1.13

diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
index 6fd624ae..7e3e4a40 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/Protocol.java
@@ -267,42 +267,50 @@ public enum Protocol
         {
             TO_CLIENT.registerPacket(
                     LoginPayloadRequest.class,
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x00 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x00 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x03 ) // Travertine - 1.13
             );
             TO_CLIENT.registerPacket(
                     Kick.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x01 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x01 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x00 ) // Travertine - 1.13
             );
             TO_CLIENT.registerPacket(
                     EncryptionRequest.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x01 ) // Travertine - 1.13
             );
             TO_CLIENT.registerPacket(
                     LoginSuccess.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x02 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x03 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x03 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x02 ) // Travertine - 1.13
             );
             TO_CLIENT.registerPacket(
                     SetCompression.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x03 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x04 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x04 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x03 ) // Travertine - 1.13
             );
 
             TO_SERVER.registerPacket(
                     LoginPayloadResponse.class,
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x00 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x00 ),
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 ) // Travertine - 1.13
             );
             TO_SERVER.registerPacket(
                     LoginRequest.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x00 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x01 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x01 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x00 ) // Travertine - 1.13
             );
             TO_SERVER.registerPacket(
                     EncryptionResponse.class,
                     map( ProtocolConstants.MINECRAFT_1_8, 0x01 ),
-                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 )
+                    map( ProtocolConstants.MINECRAFT_1_13, 0x02 ),
+                    map( ProtocolConstants.MINECRAFT_1_13_R, 0x01 ) // Travertine - 1.13
             );
         }
     };
@@ -406,6 +414,11 @@ public enum Protocol
             linkedProtocols.put( ProtocolConstants.MINECRAFT_1_12_1, Arrays.asList(
                     ProtocolConstants.MINECRAFT_1_12_2
             ) );
+            // Travertine start - 1.13
+            linkedProtocols.put (ProtocolConstants.MINECRAFT_1_13, Arrays.asList(
+                    ProtocolConstants.MINECRAFT_1_13_R
+            ));
+            // Travertine end - 1.13
         }
 
         @Getter
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index 80438990..292fece6 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -20,6 +20,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_12_1 = 338;
     public static final int MINECRAFT_1_12_2 = 340;
     public static final int MINECRAFT_1_13 = 389;
+    public static final int MINECRAFT_1_13_R = 393;
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.7.x",
             "1.8.x",
@@ -43,7 +44,8 @@ public class ProtocolConstants
             ProtocolConstants.MINECRAFT_1_12,
             ProtocolConstants.MINECRAFT_1_12_1,
             ProtocolConstants.MINECRAFT_1_12_2,
-            ProtocolConstants.MINECRAFT_1_13
+            ProtocolConstants.MINECRAFT_1_13,
+            ProtocolConstants.MINECRAFT_1_13_R
     );
 
     public static final boolean isBeforeOrEq(int before, int other)
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
index 45716724..1e5f223f 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
@@ -52,6 +52,7 @@ public abstract class EntityMap
             case ProtocolConstants.MINECRAFT_1_12_2:
                 return EntityMap_1_12_1.INSTANCE;
             case ProtocolConstants.MINECRAFT_1_13:
+            case ProtocolConstants.MINECRAFT_1_13_R:
                 return EntityMap_1_13.INSTANCE;
         }
         throw new RuntimeException( "Version " + version + " has no entity map" );
-- 
2.18.0

