From 13fb2edfae12310bd656afc2a67208c856de6b1b Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:03:12 +0800
Subject: [PATCH] Remove "Quitting" reason


diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index 00749538..7b99c466 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -269,7 +269,8 @@ public class ServerConnector extends PacketHandler
                 user.getSentBossBars().clear();
 
                 user.unsafe().sendPacket( new Respawn( login.getDimension(), login.getWorldName(), login.getSeed(), login.getDifficulty(), login.getGameMode(), login.getPreviousGameMode(), login.getLevelType(), login.isDebug(), login.isFlat(), false ) );
-                user.getServer().disconnect( "Quitting" );
+                // FlameCord - Remove "Quitting" reason
+                user.getServer().disconnect();
             } else
             {
                 // FlameCord start
@@ -364,13 +365,15 @@ public class ServerConnector extends PacketHandler
             user.setDimension( login.getDimension() );
 
             // Remove from old servers
-            user.getServer().disconnect( "Quitting" );
+            // FlameCord - Remove "Quitting" reason
+            user.getServer().disconnect();
         }
 
         // TODO: Fix this?
         if ( !user.isActive() )
         {
-            server.disconnect( "Quitting" );
+            // FlameCord - Remove "Quitting" reason
+            server.disconnect();
             // Silly server admins see stack trace and die
             bungee.getLogger().warning( "No client connected for pending server!" );
             return;
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 96df0175..fc1798fb 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -428,7 +428,8 @@ public final class UserConnection implements ProxiedPlayer
             if ( server != null )
             {
                 server.setObsolete( true );
-                server.disconnect( "Quitting" );
+                // FlameCord - Remove "Quitting" reason
+                server.disconnect();
             }
         }
     }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index ad3bdee5..c05714ba 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -92,6 +92,7 @@ public class UpstreamBridge extends PacketHandler
                 }
                 // FlameCord end
             }
+            // FlameCord - Remove "Quitting" reason
             con.getServer().disconnect( "Quitting" );
         }
     }
-- 
2.31.1

