From 36e054d017b32104a878faed79fb18b5d8b5c9df Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:14:02 +0800
Subject: [PATCH] Change default timeout


diff --git a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
index 85d2197a..3b5ce41e 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
@@ -35,7 +35,8 @@ public abstract class Configuration implements ProxyConfig
     /**
      * Time before users are disconnected due to no network activity.
      */
-    private int timeout = 30000;
+    // FlameCord - Modify default timeout
+    private int timeout = 17000;
     /**
      * UUID used for metrics.
      */
-- 
2.32.0

