From f790d57423c57a21d642db20b6de9d3515fd3e93 Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:14:02 +0800
Subject: [PATCH] Change default timeout


diff --git a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
index 8911bee85..b67195ec8 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
@@ -35,7 +35,8 @@ public abstract class Configuration implements ProxyConfig
     /**
      * Time before users are disconnected due to no network activity.
      */
-    private int timeout = 30000;
+    // FlameCord - Modify default timeout
+    private int timeout = 17000;
     /**
      * UUID used for metrics.
      */
-- 
2.37.3.windows.1

