From 1110d8e2d16633d02c7839b38f800ee5cedc0a84 Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:25:51 +0800
Subject: [PATCH] Don't allow name containing dot


diff --git a/proxy/src/main/java/net/md_5/bungee/util/AllowedCharacters.java b/proxy/src/main/java/net/md_5/bungee/util/AllowedCharacters.java
index d1cd1090..44cb1ce6 100644
--- a/proxy/src/main/java/net/md_5/bungee/util/AllowedCharacters.java
+++ b/proxy/src/main/java/net/md_5/bungee/util/AllowedCharacters.java
@@ -21,7 +21,8 @@ public final class AllowedCharacters
         } else
         {
             // Don't allow spaces, Yaml config doesn't support them
-            return isChatAllowedCharacter( c ) && c != ' ';
+            // FlameCord - Don't allow dots
+            return isChatAllowedCharacter( c ) && c != ' ' && c != '.';
         }
     }
 
-- 
2.32.0

