From 799190416ca187a0f7021813a617a6031573710b Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:35:33 +0800
Subject: [PATCH] Allow custom uuids even if onlineMode is true


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index ac20d5ad..706fb321 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -779,7 +779,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     public void setUniqueId(UUID uuid)
     {
         Preconditions.checkState( thisState == State.USERNAME, "Can only set uuid while state is username" );
-        Preconditions.checkState( !onlineMode, "Can only set uuid when online mode is false" );
+        // FlameCord - Allow custom uuids even if onlineMode is true
         this.uniqueId = uuid;
     }
 
-- 
2.40.1.windows.1

