From d3de5793409f78a1c29503afe900c04ce69f6b4d Mon Sep 17 00:00:00 2001
From: foss-mc <69294560+foss-mc@users.noreply.github.com>
Date: Wed, 16 Dec 2020 18:40:55 +0800
Subject: [PATCH] Change IllegalStateException to QuietException and explain
 that is a plugin


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 43042c6a..9b04f7ca 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -308,7 +308,8 @@ public final class UserConnection implements ProxiedPlayer
 
             if ( getServer() == null && !ch.isClosing() )
             {
-                throw new IllegalStateException( "Cancelled ServerConnectEvent with no server or disconnect." );
+                // FlameCord - Change IllegalStateException to QuietException and explain that is a plugin
+                throw new QuietException("A plugin cancelled ServerConnectEvent with no server or disconnect.");
             }
             return;
         }
-- 
2.32.0.windows.1

