From 337765289b483876d466a7a194492a625f75e2a7 Mon Sep 17 00:00:00 2001
From: xIsm4 <minelatinsoporte@gmail.com>
Date: Sun, 17 Jul 2022 11:13:51 +0200
Subject: [PATCH] Fix firewall system


diff --git a/flamecord/src/main/java/dev/_2lstudios/flamecord/configuration/FlameCordConfiguration.java b/flamecord/src/main/java/dev/_2lstudios/flamecord/configuration/FlameCordConfiguration.java
index 31631626..c443b8d1 100644
--- a/flamecord/src/main/java/dev/_2lstudios/flamecord/configuration/FlameCordConfiguration.java
+++ b/flamecord/src/main/java/dev/_2lstudios/flamecord/configuration/FlameCordConfiguration.java
@@ -36,6 +36,8 @@ public class FlameCordConfiguration extends FlameConfig {
     @Getter
     private Collection<String> antibotCountryBlacklist = Arrays.asList("CN", "HK", "RU", "IN", "TH", "ID", "DZ", "VN", "IR", "PK");
     @Getter
+    private Collection<String> antibotFirewalledPackets = Arrays.asList("BadPacketException","QuietException","IllegalStateConfig","FastException");
+    @Getter
     private boolean antibotCountryLog = true;
     @Getter
     private boolean antibotFastChatEnabled = true;
@@ -108,6 +110,7 @@ public class FlameCordConfiguration extends FlameConfig {
         this.antibotFastChatTime = setIfUnexistant("antibot.fastchat.time", this.antibotFastChatTime, config);
         this.antibotFastChatLog = setIfUnexistant("antibot.fastchat.log", this.antibotFastChatLog, config);
         this.antibotFirewallEnabled = setIfUnexistant("antibot.firewall.enabled", this.antibotFirewallEnabled, config);
+        this.antibotFirewalledPackets = setIfUnexistant("antibot.firewall.packets", this.antibotFirewalledPackets, config);
         this.antibotFirewallIpset = setIfUnexistant("antibot.firewall.ipset", this.antibotFirewallIpset, config);
         this.antibotFirewallExpire = setIfUnexistant("antibot.firewall.time", this.antibotFirewallExpire, config);
         this.antibotFirewallLog = setIfUnexistant("antibot.firewall.log", this.antibotFirewallLog, config);
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
index dc2b6a7f..68d7b732 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
@@ -155,7 +155,7 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
             boolean logExceptions = !( handler instanceof PingHandler );
 
             // Flamecord start - Antibot System
-            if (!(cause instanceof ReadTimeoutException) && !(cause instanceof IOException))
+            if (cause == FlameCord.getInstance().getFlameCordConfiguration().getAntibotFirewalledPackets())
             {
                 FlameCord.getInstance().getAddressDataManager().getAddressData(ctx.channel().remoteAddress()).firewall();
                 BungeeCord.getInstance().getLogger().log( Level.INFO, "[FlameCord] [{0}] was firewalled because of " + cause.getClass().getName(), ctx.channel().remoteAddress() );
-- 
2.34.1.windows.1

