From 5e0a6aa349884b190a106b2c7609d6482ee5360f Mon Sep 17 00:00:00 2001
From: LinsaFTW <25271111+linsaftw@users.noreply.github.com>
Date: Tue, 15 Nov 2022 18:07:27 +0100
Subject: [PATCH] Show connection fail reason


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index f3d60253..999f02e5 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -406,7 +406,8 @@ public final class UserConnection implements ProxiedPlayer
         if(FlameCord.getInstance().getFlameCordConfiguration().isLoggerDetailedConnection()) {
             bungee.getLogger().log(Level.WARNING, "Error occurred processing connection for " + this.name + " " + Util.exception(cause, false)); // Waterfall
         }
-        return ""; // Waterfall
+        // FlameCord - Show connection fail reason
+        return cause.getMessage();
     }
 
     @Override
-- 
2.37.3.windows.1

