From efa69c69e7b81b75ea1a602f07b3410cacbcc89c Mon Sep 17 00:00:00 2001
From: LinsaFTW <25271111+linsaftw@users.noreply.github.com>
Date: Tue, 15 Nov 2022 18:07:27 +0100
Subject: [PATCH] Only show detailed connect exception to admins


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index f3d60253..713d9037 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -406,7 +406,8 @@ public final class UserConnection implements ProxiedPlayer
         if(FlameCord.getInstance().getFlameCordConfiguration().isLoggerDetailedConnection()) {
             bungee.getLogger().log(Level.WARNING, "Error occurred processing connection for " + this.name + " " + Util.exception(cause, false)); // Waterfall
         }
-        return ""; // Waterfall
+        // FlameCord - Show connection fail reason
+        return groups.contains( "admin" ) ? cause.getMessage() : cause.getClass().getName();
     }
 
     @Override
-- 
2.29.2.windows.2

